allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,275 @@
+            if self.training and all(torch.rand(1) >= self._scheduled_sampling_ratio):
+                input_choices = targets[:, timestep]
+            else:
+                if timestep == 0:
+                    # For the first timestep, when we do not have targets, we input start symbols.
+                    # (batch_size,)
+                    with torch.cuda.device_of(final_encoder_output):
+                        input_choices = Variable(torch.LongTensor([self._start_index]).expand_as(
+                                final_encoder_output[:, 0]))
+                else:
+                    input_choices = last_predictions
+            decoder_input = self._prepare_decode_step_input(input_choices, decoder_hidden, encoder_outputs)
+            decoder_hidden, decoder_context = self._decoder_cell(decoder_input,
+                                                                 (decoder_hidden, decoder_context))
+            # (batch_size, num_classes)
+            output_projections = self._output_projection_layer(decoder_hidden)
+            # (batch_size, 1, num_classes)
+            step_logits.append(output_projections.unsqueeze(1))
+            decoder_predictions = self._predict_step(output_projections)
+            step_probabilities.append(decoder_predictions["class_probabilities"].unsqueeze(1))
+            last_predictions = decoder_predictions["predicted_classes"]
+            # (batch_size, 1)
+            step_predictions.append(last_predictions.view(-1, 1))
+        # step_logits is a list containing tensors of shape (batch_size, 1, num_classes)
+        # This is (batch_size, num_decoding_steps, num_classes)
+        logits = torch.cat(step_logits, 1)
+        class_probabilities = torch.cat(step_probabilities, 1)
+        all_predictions = torch.cat(step_predictions, 1)
+        output_dict = {"logits": logits,
+                       "class_probabilities": class_probabilities,
+                       "predictions": all_predictions}
+        if target_tokens:
+            target_mask = get_text_field_mask(target_tokens)
+            # `targets` currently contains START + target_sequence + END, and logits correspond to the outputs
+            # from these timesteps. During training, we want the output from timestep i to be similar to the
+            # input token from timestep i + 1. That is, the comparison between logits and targets should be
+            # one-off.
+            # Consider a single example where the target has 3 words, and padding is to 7 tokens. The complete
+            #   sequence would correspond to <S> w1 w2 w3 <E> <P> <P>
+            #   and the mask would be         1  1  1  1   1   0   0
+            #   and let the logits be         l1 l2 l3 l4  l5  l6  l7
+            # We actually need to compare:
+            #   the sequence w1 w2 w3 <E> <P> <P>
+            #   with masks   1  1  1   1   0   0
+            #   against      l1 l2 l3  l4  l5  l6
```

Comment:
This should be `l2 l3  l4  l5  l6 l7`, I think?

Reply:
No, this is correct.  You really want to have four lines here, where the new line shows the decoder _inputs_.  This would be more clear.  The first column is then: `input: <S>, target: w1, logit: l1, mask: 1`.